src/Compiler/NodeVisitor/AddVarTypesNodeVisitor.php

```diff
@@ -67,7 +68,15 @@ public function enterNode(Node $node): ?Node
 
             $variableStatements[] = $docNop;
         }
-        $node->stmts = array_merge($variableStatements, $statements);
+
+        $variableStatements[] = new Expression(
```

Comment:
This visitor is executed multiple times, so this is not very good way to mark end of variables. Also it is added to method before header assignements so it will not help

Reply:
That code will be in final template multiple times but it does nothing so I think it is not a problem.

Point of this code is that variable type doc comments are attached to this placeholder node. So it is copied even when next statement (return) is not. I think variables are first added to prepare/main and then it is copied from there together with header assignments.

Reply:
I think it is useless. What is the reason to add it here if it does nothing?

Reply:
It hoes nothing except that it binds type doc comments. We need to bind them to some neutral node instead of random first node that follows in method body. So it is not stripped during copying and also to prevent potential errors. I encountered strange error in my project that I solved by workaround but now I see it was caused by this.

That error manifests when first statement in method body is foreach. Doc comment above foreach is allowed to define only key and value avriables of foreach otherwise PHPStan reports error.

Reply:
I think I should add test for that now

Reply:
Ah it's “doc keeper” statement like the `$__marker___ = $this;` was in my PR. Got it now

Reply:
Test added